### PR TITLE
Optimize SqlFunctionCall/SqlCustomFunctionCall

### PR DESCRIPTION
**Pull request by SergeiPavlov**
* Make abstract class `SqlFunctionCallBase` with functionality common for both `SqlFunctionCall` & `SqlCustomFunctionCall`
* Change type of  `SqlFunctionCall.Arguments` to `IReadOnlyList` to avoid unnecessary allocations. Argument list is now immutable.
* Replace `Collection<SqlExpression>` to `List<SqlExpression>`.  List is more lightweight with same functionality
* Make `ArgumentValidator.EnsureArgumentIs<T>()` to return `T` to avoid double type casting after call
* Make EnsureArgumentIs's parameter `parameterName` automatic by using `[CallerArgumentExpression]` attribute
* Remove invocation of `EnsureArgumentNotNull` before `EnsureArgumentIs` because the later checks for `null` as well.

### CODE REVIEW

**Review by alex-kulakov — CHANGES_REQUESTED** [review]

**Comment by alex-kulakov** [comment]
Also worth mentioning that FunctionCalls lost ability to change parameters after creation.

File: Orm/Xtensive.Orm/Sql/Dml/Expressions/SqlNextValue.cs [hunk 1]
```diff
@@ -37,9 +37,7 @@ public Sequence Sequence
 
     public override void ReplaceWith(SqlExpression expression)
     {
-      ArgumentValidator.EnsureArgumentNotNull(expression, "expression");
-      ArgumentValidator.EnsureArgumentIs<SqlNextValue>(expression, "expression");
-      SqlNextValue replacingExpression = expression as SqlNextValue;
+      SqlNextValue replacingExpression = ArgumentValidator.EnsureArgumentIs<SqlNextValue>(expression);
```
[the remark this answers is not in the document]
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Sql/Dml/Expressions/SqlNextValue.cs]: done

File: Orm/Xtensive.Orm/Sql/Dml/Expressions/SqlUserColumn.cs [hunk 2]
```diff
@@ -23,9 +23,7 @@ public SqlExpression Expression
 
     public override void ReplaceWith(SqlExpression expression)
     {
-      ArgumentValidator.EnsureArgumentNotNull(expression, "expression");
-      ArgumentValidator.EnsureArgumentIs<SqlUserColumn>(expression, "expression");
-      SqlUserColumn replacingExpression = expression as SqlUserColumn;
+      SqlUserColumn replacingExpression = ArgumentValidator.EnsureArgumentIs<SqlUserColumn>(expression);
```
[the remark this answers is not in the document]
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Sql/Dml/Expressions/SqlUserColumn.cs]: done

File: Orm/Xtensive.Orm/Sql/Dml/Expressions/SqlNull.cs [hunk 3]
```diff
@@ -13,7 +13,7 @@ public class SqlNull : SqlExpression
     public override void ReplaceWith(SqlExpression expression)
     {
       ArgumentValidator.EnsureArgumentNotNull(expression, "expression");
-      ArgumentValidator.EnsureArgumentIs<SqlNull>(expression, "expression");
+      ArgumentValidator.EnsureArgumentIs<SqlNull>(expression);
```
[the remark this answers is not in the document]
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Sql/Dml/Expressions/SqlNull.cs]: sure, done

File: Orm/Xtensive.Orm/Core/ArgumentValidator.cs [hunk 4]
```diff
@@ -92,12 +92,14 @@ public static void EnsureArgumentNotNullOrEmptyOrWhiteSpace(string value, [Invok
     /// <param name="parameterName">Name of the method parameter.</param>
     /// <typeparam name="T">The expected type of value.</typeparam>
     [MethodImpl(MethodImplOptions.AggressiveInlining)]
-    public static void EnsureArgumentIs<T>(object value, [InvokerParameterName] string parameterName)
+    public static T EnsureArgumentIs<T>(object value,
```
[the remark this answers is not in the document]
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Core/ArgumentValidator.cs:96]: added the XML doc line

File: Orm/Xtensive.Orm/Sql/Dml/Expressions/SqlVariable.cs [hunk 5]
```diff
@@ -38,9 +38,7 @@ public SqlDeclareVariable Declare()
 
     public override void ReplaceWith(SqlExpression expression)
     {
-      ArgumentValidator.EnsureArgumentNotNull(expression, "expression");
-      ArgumentValidator.EnsureArgumentIs<SqlVariable>(expression, "expression");
-      SqlVariable replacingExpression = expression as SqlVariable;
+      SqlVariable replacingExpression = ArgumentValidator.EnsureArgumentIs<SqlVariable>(expression);
```
[the remark this answers is not in the document]
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Sql/Dml/Expressions/SqlVariable.cs]: done

File: Orm/Xtensive.Orm/Sql/SqlDml.cs [hunk 6]
```diff
@@ -738,15 +738,13 @@ public static SqlUserFunctionCall FunctionCall(string name, IEnumerable<SqlExpre
...
-        collection.Add(expressions[i]);
-      return FunctionCall(name, collection);
+      SqlValidator.EnsureAreSqlRowArguments(expressions);
+      return new SqlUserFunctionCall(name, expressions);
     }
 
     public static SqlUserFunctionCall FunctionCall(string name)
     {
-      return FunctionCall(name, new Collection<SqlExpression>());
+      return FunctionCall(name, System.Array.Empty<SqlExpression>());
```
[the remark this answers is not in the document]
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Sql/SqlDml.cs]: done

**Comment by SergeiPavlov** [comment]
> Also worth mentioning that FunctionCalls lost ability to change parameters after creation.

I mentioned  in the PR description that `Arguments` is now immutable.

**Comment by alex-kulakov** [comment]
Yes, you did, pardon me for inattention.

**Review by alex-kulakov — APPROVED** [review]

**Pull request merged**